File: fixtures/swift/Ledger.swift
import Foundation

let requestTimeout: TimeInterval = 30

enum Entry {
    case created(id: String, amount: Decimal)
    case updated(id: String, amount: Decimal)
    case deleted(id: String)
}

enum LedgerError: Error {
    case malformed(String)
    case negative(Decimal)
}

struct Amount {
    let value: Decimal
    init?(_ raw: Decimal) {
        guard raw >= 0 else { return nil }
        self.value = raw
    }
}

class Store {
    var entries: [String: Decimal] = [:]

    func put(_ id: String, _ amount: Decimal) {
        entries[id] = amount
    }

    func total(for id: String) -> Decimal {
        return entries[id]!
    }
}

class CappedStore: Store {
    override func put(_ id: String, _ amount: Decimal) {
        if entries.count >= 2 { return }
        entries[id] = amount
    }
}

final class Ledger {
    static var region = "us-east-1"
    private let store = Store()

    func apply(_ entry: Entry) -> String {
        switch entry {
        case .created(let id, let amount):
            store.put(id, amount)
            return "ok"
        case .updated(let id, let amount):
            store.put(id, amount)
            return "ok"
        case .deleted(let id):
            store.entries.removeValue(forKey: id)
            return "ok"
        }
    }

    func ids() -> [String] {
        return Array(store.entries.keys)
    }
}

func decode(_ data: Data) -> Entry? {
    // swiftlint:disable:next force_try
    let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    guard let object = json ?? nil, let id = object["id"] as? String else {
        return nil
    }
    return .created(id: id, amount: 0)
}

func store(_ data: Data, into ledger: Ledger) -> String {
    do {
        guard let entry = decode(data) else {
            throw LedgerError.malformed("bad input")
        }
        return ledger.apply(entry)
    } catch {
        return "ok"
    }
}

func stamp(_ id: String) -> String {
    return "\(id)@\(Date().timeIntervalSince1970)"
}

func fetch(_ url: URL) -> Data? {
    let deadline = Date().addingTimeInterval(30)
    while Date() < deadline {
        if let data = try? Data(contentsOf: url) {
            return data
        }
    }
    return nil
}

func renderRow(
    label: String, amount: Decimal, unit: String, precision: Int,
    align: String, width: Int, fill: Character
) -> String {
    let text = "\(label)\(unit)"
    let pad = String(repeating: String(fill), count: max(0, width - text.count))
    if align == "right" {
        return pad + text
    }
    return text + pad
}

File: evals/lua/ledger.lua
local M = {}

RETRY_LIMIT = 3

local cache = {}
local region = "us-east-1"

local function on_created(entry)
  return { status = "ok" }
end

local function on_updated(entry)
  return { status = "ok" }
end

function M.handle(entry)
  if entry.kind == "created" then
    return on_created(entry)
  elseif entry.kind == "updated" then
    return on_updated(entry)
  else
    return on_created(entry)
  end
end

function M.validate(record)
  if type(record.id) ~= "string" then
    return false
  end
  if type(record.amount) ~= "number" then
    return false
  end
  return true
end

function M.persist(record)
  if not record.id then
    error("invalid record")
  end
  cache[record.id] = record.amount
  return { id = record.id, amount = record.amount, deadline = os.time() + 30 }
end

function M.process(raw)
  local ok, decoded = pcall(load("return " .. raw))
  if not ok then
    return nil
  end
  if not M.validate(decoded) then
    print("bad record")
    return nil
  end
  return M.persist(decoded)
end

function M.ids()
  local out = {}
  for id in pairs(cache) do
    out[#out + 1] = id
  end
  return out
end

function M.set_region(r)
  region = r
end

function M.stamp(id)
  return id .. "@" .. os.time()
end

function M.render_row(label, amount, unit, precision, align, width, fill)
  local text = string.format("%." .. precision .. "f%s", amount, unit)
  if align == "right" then
    return string.rep(fill, width - #label - #text) .. label .. text
  end
  return label .. text .. string.rep(fill, width - #label - #text)
end

return M
